### PR TITLE
Refactor the entire library

### PR DESCRIPTION
**Pull request by squimrel**
Compiles and runs. Should have the same behavior as before.
Function signatures of exposed functions were not modified.

### CODE REVIEW

File: libcheckisomd5.c [hunk 1]
```diff
@@ -32,336 +32,111 @@
...
 
-        nread = read(isofd, buf, nattempt);
-        if (nread <= 0)
+        ssize_t nread = read(isofd, buffer, nbyte);
+        if (nread <= 0L)
             break;
 
-        if (nread > nattempt) {
-            nread = nattempt;
-            lseek(isofd, offset+nread, SEEK_SET);
```
Comment (squimrel) [inline libcheckisomd5.c]: This should never happen this check was added back in 2005 in 1b993e6ac6db0f6cc5cb897702350d23331b057f. According the author @vathpela some weird behavior occurred where it returned a value greater than `nattempt`. I'll just take a leap of faith here by assuming that this was fixed in the past 12 years.

File: libcheckisomd5.c [hunk 2]
```diff
@@ -32,336 +32,111 @@
...
-    if (cb)
-        cb(cbdata, 0, isosize - skipsectors*2048);
+    MD5_CTX hashctx;
+    MD5_Init(&hashctx);
 
-    while (offset < isosize - skipsectors*2048) {
-        nattempt = MIN(isosize - skipsectors*2048 - offset, bufsize);
+    const size_t buffer_size = NUM_SYSTEM_SECTORS * SECTOR_SIZE;
+    unsigned char *buffer;
+    buffer = memalign((size_t) getpagesize(), buffer_size * sizeof(*buffer));
```
Comment (squimrel) [inline libcheckisomd5.c]: Generally speaking the page-aligned buffer might speed thinks up but also it's for example required by udisks so it's essential for some people who want to use this library.

File: libcheckisomd5.c [hunk 3]
```diff
@@ -32,336 +32,111 @@
...
-        offset += 2048L;
+#include "utilities.h"
+
+static void clear_appdata(unsigned char *const buffer, const size_t size, const off_t appdata_offset, const off_t offset) {
+    static const ssize_t buffer_start = 0;
+    const ssize_t difference = appdata_offset - offset;
+    if (buffer_start <= difference && difference <= size) {
+        const size_t clear_start = (size_t) MAX(buffer_start, difference);
+        const size_t clear_len = MIN(size, (size_t)(difference + APPDATA_SIZE)) - clear_start;
+        memset(buffer + clear_start, ' ', clear_len);
```
Comment (squimrel) [inline libcheckisomd5.c]: I hope I've done this right...? Previous implementation can be found here: [`5992e7`](https://github.com/rhinstaller/isomd5sum/commit/5992e7fb80886b2bfe2be3b7c2029bdd767fddb0#diff-6a1a1d0377babbfeecf61ed692d67f50R110).

**Comment by squimrel** [comment]
@linux-modder @MartinBriza @bcl

File: libcheckisomd5.c (hunk 1 above)
Comment (MartinBriza) [inline libcheckisomd5.c]: This is a pretty weird check. However, I think it's harmless and could be just left there with some small comment on the reasons it was added. The performance hit of such a check (especially if accompanied by a compiler hint) is not great.

File: libcheckisomd5.c (hunk 2 above)
Comment (MartinBriza) [inline libcheckisomd5.c]: I'll just specify what "required by udisks" means: There is a OpenForBenchmark call provided by UDisks2 which provides a FD that's opened with the O_DIRECT flag and therefore requires you to work with page-aligned buffers. This call is then (ab)used by Fedora Media Writer for writing and checking if the image was written correctly.

File: libcheckisomd5.c [hunk 4]
```diff
@@ -21,7 +21,7 @@
  * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
  */
 
-
+#include <malloc.h>
```
Comment (MartinBriza) [inline libcheckisomd5.c]: malloc.h? why? stdlib is not enough?

Comment (squimrel) [inline libcheckisomd5.c]: Because `memalign` is provided by `malloc.h` on most systems. But looking at this again it seems like `alligned_alloc` is probably the better choice since it's even c11 standard.

File: libcheckisomd5.c (hunk 1 above)
Comment (squimrel) [inline libcheckisomd5.c]: But there're other places where `read` is used in this library. Should this check be added in all places then? Also I'm afraid that it'll just confuse other developers who'll look at it. So if it lands it definitely has to land in it's own commit with a lot of explanation.

**Comment by MartinBriza** [comment]
From my point of view, this looks good. Code is a lot cleaner, and separated better.
I guess my only concern was about removing the (seemingly unnecessary) check which may be still valid for some platforms.
Anyway, I'll try to compile and test the code on Windows and Mac at some point of the week and report back to this PR.

File: libcheckisomd5.c (hunk 1 above)
Comment (squimrel) [inline libcheckisomd5.c]: I did add the check back in. I'm sorry that I was so stubborn about it but it totally makes sense to just leave it.

File: libimplantisomd5.c [hunk 5]
```diff
@@ -31,240 +31,139 @@
...
-    return offset;
-}
-
-
-static unsigned int writeAppData(unsigned char *appdata, char *valstr, unsigned int loc) {
-    if (loc + strlen(valstr) > 511) {
+static size_t writeAppData(unsigned char *const appdata, const char *const valstr, const size_t loc) {
+    if (loc + strlen(valstr) >= APPDATA_SIZE) {
         printf("Attempted to write too much appdata, exiting...\n");
         exit(-1);
```
Comment (bcl) [inline libimplantisomd5.c]: This exit should be removed and the callers should check for failure. libraries really shouldn't be calling exit, especially when we have a decent way to return error messages 1 layer above.

File: libimplantisomd5.c [hunk 6]
```diff
@@ -31,240 +31,139 @@
...
         *errstr = "Error - Unable to open file %s\n\n";
         return -1;
     }
 
-    pvd_offset = parsepvd(isofd, mediasum, &isosize);
-    if (pvd_offset < 0) {
+    off_t pvd_offset;
+    const off_t isosize = primary_volume_size(isofd, &pvd_offset);
+    if (isosize == 0) {
         *errstr = "Could not find primary volumne!\n\n";
```
Comment (bcl) [inline libimplantisomd5.c]: May as well fix this previously existing typo - 'primary volume'

File: libimplantisomd5.c [hunk 7]
```diff
@@ -31,240 +31,139 @@
...
         return -1;
     }
 
     lseek(isofd, pvd_offset + APPDATA_OFFSET, SEEK_SET);
-    nread = read(isofd, orig_appdata, 512);
+    unsigned char appdata[APPDATA_SIZE];
+    read(isofd, appdata, APPDATA_SIZE);
```
Comment (bcl) [inline libimplantisomd5.c]: Need to check the return value of read and return an error.

File: libimplantisomd5.c [hunk 8]
```diff
@@ -31,240 +31,139 @@
...
+    if (lseek(isofd, pvd_offset + APPDATA_OFFSET, SEEK_SET) < 0)
         printf("seek failed\n");
 
-    i = write(isofd, new_appdata, 512);
-    if (i<0) {
-        printf("write failed %d\n", i);
+    ssize_t error = write(isofd, appdata, APPDATA_SIZE);
+    if (error < 0) {
+        printf("write failed %ld\n", error);
         perror("");
```
Comment (bcl) [inline libimplantisomd5.c]: I think this should return an error to the caller instead of just a perror.

File: libimplantisomd5.c [hunk 9]
```diff
@@ -31,240 +31,139 @@
...
+    loc = writeAppData(appdata, ";", loc);
 
-    loc = writeAppData(new_appdata, "THIS IS NOT THE SAME AS RUNNING MD5SUM ON THIS ISO!!", loc);
+    loc = writeAppData(
+        appdata, "THIS IS NOT THE SAME AS RUNNING MD5SUM ON THIS ISO!!", loc);
 
-    i = lseek(isofd, pvd_offset + APPDATA_OFFSET, SEEK_SET);
-    if (i<0)
+    if (lseek(isofd, pvd_offset + APPDATA_OFFSET, SEEK_SET) < 0)
         printf("seek failed\n");
```
Comment (bcl) [inline libimplantisomd5.c]: This should return a error to the caller, unless you can think of any cases where the seek can fail but the subsequent write is ok.

File: utilities.c [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+    }
+    return true;
+}
+
+void md5sum(char *const hashsum, MD5_CTX *const hashctx) {
+    unsigned char digest[HASH_SIZE / 2];
+    MD5_Final(digest, hashctx);
+    *hashsum = '\0';
+    for (size_t i = 0; i < HASH_SIZE / 2; i++) {
+        char tmp[3];
```
Comment (squimrel) [inline utilities.c]: @bcl regarding the `snprintf` comment above. Why is that? Should it be four here then because of `"%02x"`?

File: libcheckisomd5.c [hunk 11]
```diff
@@ -32,336 +31,120 @@
...
-        else if (buf[0] == 255)
-            /* hit end and didn't find primary volume descriptor */
-            return ((long long)-1);
-        offset += 2048L;
+#include "utilities.h"
+
+static void clear_appdata(unsigned char *const buffer, const size_t size, const off_t appdata_offset, const off_t offset) {
+    static const ssize_t buffer_start = 0;
+    const ssize_t difference = appdata_offset - offset;
+    if (-APPDATA_SIZE <= difference && difference <= (ssize_t) size) {
```
Comment (squimrel) [inline libcheckisomd5.c]: Ehm because I don't get how I'd do that. :/

File: checkisomd5.c [hunk 12]
```diff
@@ -115,73 +110,71 @@ int processExitStatus(int rc) {
...
-    optCon = poptGetContext("checkisomd5", argc, (const char **)argv, options, 0);
+    poptContext optCon = poptGetContext("checkisomd5", argc, (const char **) argv, options, 0);
 
-    if ((rc = poptGetNextOpt(optCon)) < -1) {
+    int rc = poptGetNextOpt(optCon);
+    if (rc < -1) {
         fprintf(stderr, "bad option %s: %s\n",
                 poptBadOption(optCon, POPT_BADOPTION_NOALIAS),
                 poptStrerror(rc));
-        exit(1);
```
Comment (squimrel) [inline checkisomd5.c]: Ehh yeah I'm not good at fixing memory leaks. o_O. Fixed :-)

File: utilities.c [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+
+bool validate_fragment(const MD5_CTX *const hashctx, const size_t fragment,
+                       const size_t fragmentsize, const char *const fragmentsums, char *hashsums) {
+    unsigned char digest[HASH_SIZE / 2];
+    MD5_CTX ctx;
+    memcpy(&ctx, hashctx, sizeof(ctx));
+    MD5_Final(digest, &ctx);
+    size_t j = (fragment - 1) * fragmentsize;
+    for (size_t i = 0; i < MIN(fragmentsize, HASH_SIZE / 2); i++) {
+        char tmp[2];
```
Comment (squimrel) [inline utilities.c]: In addition to the comment on this below for better understanding: I thought that I'd only need two because I need one for `%01x` and one for `\0`.

File: utilities.c (hunk 10 above)
Comment (bcl) [inline utilities.c]: It needs to be at least 3 because the output is 2 bytes + 0x00, and snprintf won't stop at 1 byte.
When I run the build in koji it has different build options and it complains that the output won't be truncated. I'm not sure why the old code did it that way, but it shouldn't matter since the following copy grabs 1 byte no matter what.

File: libcheckisomd5.c (hunk 11 above)
Comment (bcl) [inline libcheckisomd5.c]: 'if (difference >= APPDATA_SIZE && difference <= (ssize_t) size)'
(or maybe I'm being dense? totally possible at this time of the morning)

File: utilities.c (hunk 13 above)
Comment (squimrel) [inline utilities.c]: I still don't see how this needs to be 3.

man snprintf:
> The functions snprintf() and vsnprintf() write at most size bytes (including the terminating null byte ('\0')) to str.

Therefore it's guaranteed that only 2 bytes will be written.
Why do you think that output would be truncated? There are two bytes one for `%01x` and on for `\0` (as mentioned in my comment above).

File: utilities.c (hunk 10 above)
Comment (squimrel) [inline utilities.c]: I'm confused let's move back to to the review comment thread above on the same topic.

File: libcheckisomd5.c (hunk 11 above)
Comment (squimrel) [inline libcheckisomd5.c]: `difference` might be negative.

Therefore if I'd flip it I would still have to negate `APPDATA_SIZE`: `difference >= -APPDATA_SIZE`.

I wrote it that way because when I developed this at my whiteboard it said: `-APPDATA_SIZE <= difference <= size` and I thought it's more clear like that in code since it indicates that difference should be in between those two values.

Comment (bcl) [inline libcheckisomd5.c]: Ah, ok. That makes sense.

File: utilities.c (hunk 13 above)
Comment (bcl) [inline utilities.c]: I think the relevant part of the manpage is from the `Field Width` section: "In no case does a nonexistent or small field width cause truncation of a field;" so `%01` will not output 1, it will output 2 bytes. And gcc in a mockbuild complains about it like this:
```
utilities.c: In function 'validate_fragment':
utilities.c:194:26: error: '__builtin___snprintf_chk' output may be truncated before the last format character [-Werror=format-truncation=]
         snprintf(tmp, 2, "%01x", digest[i]);
                          ^~~~~~
In file included from /usr/include/stdio.h:889:0,
                 from utilities.c:19:
/usr/include/bits/stdio2.h:64:10: note: '__builtin___snprintf_chk' output between 2 and 3 bytes into a destination of size 2
   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        __bos (__s), __fmt, __va_arg_pack ());
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Comment (squimrel) [inline utilities.c]: That's what I missed. Thanks for being so patient with me.

File: utilities.c (hunk 10 above)
Comment (squimrel) [inline utilities.c]: Thanks for clarification.
